tokio-trace-futures/examples/proxy_server.rs

```diff
@@ -0,0 +1,153 @@
+                    info!(
+                        message = "client connected",
+                        client_addr = field::display(&client_addr)
+                    )
+                }
+                Err(e) => debug!(
+                    message = "could not get client info",
+                    error = field::display(e)
+                ),
+            }
+
+            let amounts = server.and_then(move |server| {
+                // Create separate read/write handles for the TCP clients that we're
+                // proxying data between. Note that typically you'd use
+                // `AsyncRead::split` for this operation, but we want our writer
+                // handles to have a custom implementation of `shutdown` which
+                // actually calls `TcpStream::shutdown` to ensure that EOF is
+                // transmitted properly across the proxied connection.
+                //
+                // As a result, we wrap up our client/server manually in arcs and
+                // use the impls below on our custom `MyTcpStream` type.
+                let client_reader = MyTcpStream(Arc::new(Mutex::new(client)));
+                let client_writer = client_reader.clone();
+                let server_reader = MyTcpStream(Arc::new(Mutex::new(server)));
+                let server_writer = server_reader.clone();
+
+                // Copy the data (in parallel) between the client and the server.
+                // After the copy is done we indicate to the remote side that we've
+                // finished by shutting down the connection.
+                let client_to_server = copy(client_reader, server_writer)
+                    .and_then(|(n, _, server_writer)| {
+                        info!(size = n);
+                        shutdown(server_writer).map(move |_| n)
+                    })
+                    .instrument(span!("client_to_server"));
+
+                let server_to_client = copy(server_reader, client_writer)
+                    .and_then(|(n, _, client_writer)| {
+                        info!(size = n);
+                        shutdown(client_writer).map(move |_| n)
+                    })
+                    .instrument(span!("server_to_client"));
+
+                client_to_server.join(server_to_client)
+            });
+
+            let msg = amounts
+                .map(move |(from_client, from_server)| {
+                    info!(
+                        client_to_server = from_client,
+                        server_to_client = from_server
+                    );
+                })
+                .map_err(|e| {
+                    // Don't panic. Maybe the client just disconnected too soon.
+                    debug!(error = field::display(e));
+                })
+                .instrument(span!(
+                    "transfer",
+                    client_address = field::debug(&client_addr),
+                    server_address = field::debug(&server_addr.to_string())
```

Comment:
Again, the `to_string` call isn't really necessary here --- you get the same output with just
```suggestion
                    server_address = field::display(&server_addr)
```

Reply:
Done